File: semantic_kernel/skill_definition/sk_function_decorator.py
# Copyright (c) Microsoft. All rights reserved.

from functools import wraps


def sk_function(name: str):
    """
    Decorator for SK functions.

    Args:
        name -- The name of the function
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            return func(*args, **kwargs)

        wrapper.__sk_function_name__ = name
        return wrapper

    return decorator

File: typescript/src/registry/skFunctionDecorator.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { ISKMethodInfo } from './iSKMethodInfo';

/**
 * Decorator required to register native functions into the kernel.
 * @remarks
 * The registration is required by the prompt templating engine and by the pipeline generator (aka planner).
 * The quality of the description affects the planner ability to reason about complex tasks.
 * The description is used both with LLM prompts and embedding comparisons.
 */
export function SKFunction(
    description: string
): (target: any, propertyKey: string, descriptor: PropertyDescriptor) => void {
    return (target: any, propertyKey: string, descriptor: PropertyDescriptor) => {
        if (typeof descriptor.value !== 'function') {
            throw new Error(`@SKFunction(${description}) decorator can only be applied to methods.`);
        }

        const info = descriptor.value as ISKMethodInfo;
        info.hasSkFunctionAttribute = true;
        info.description = description;
        if (!Array.isArray(info.parameters)) {
            info.parameters = [];
        }
    };
}
